envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/synchronization/GlobalServiceChanges.kt

```diff
@@ -34,4 +40,33 @@ class GlobalServiceChanges(
             .checkpoint("global-service-changes-emitted")
             .name("global-service-changes-emitted").metrics()
     }
+
+    private fun combinedExperimentalFlow(
+        serviceStatesStreams: List<Flux<Set<LocalityAwareServicesState>>>
+    ): Flux<List<LocalityAwareServicesState>> {
+
+        return Flux.combineLatest(
+            serviceStatesStreams.map {
+                // if a number of items emitted by one source is very high, combineLatest may entirely ignore items
+                // emitted by other sources. publishOn with multithreaded scheduler prevents it.
+                //
+                // onBackpressureLatest ensures that combineLatest sources respect backpressure, so combineLatest
+                // will not fail with "Queue is full: Reactive Streams source doesn't respect backpressure" suppressed
+                // error
+                it.publishOn(scheduler, 1).onBackpressureLatest()
```

Comment:
> Now it might create a lot of threads.
No, it creates 3 threads now instead of 2 before. This is not a parallel operation, so every `publishOn`, etc. only need one thread.

Reply:
And this particular `publishOn` is not changed, it is already here in "normal" flow. I added another `publishOn` in line 68